File: lib/csrc/nn/topk_pwd.cpp
#include <torch/extension.h>

#include <vector>

// CUDA forward declarations

void topk_pwd_forward_cuda(const torch::Tensor vid,
    const torch::Tensor inds0, const torch::Tensor inds1,
    torch::Tensor dists, int ps, int pt,
    int dilation, bool reflect_bounds, bool use_adj,
    int off_H0, int off_W0, int off_H1, int off_W1);

// C++ interface

#define CHECK_CUDA(x) TORCH_CHECK(x.device().is_cuda(), #x " must be a CUDA tensor")
#define CHECK_CONTIGUOUS(x) TORCH_CHECK(x.is_contiguous(), #x " must be contiguous")
#define CHECK_INPUT(x) CHECK_CUDA(x); CHECK_CONTIGUOUS(x)

void topk_pwd_forward(const torch::Tensor vid,
                      const torch::Tensor inds0, const torch::Tensor inds1,
                      torch::Tensor dists, int ps, int pt,
                      int dilation, bool reflect_bounds, bool use_adj,
                      int off_H0, int off_W0, int off_H1, int off_W1){
  CHECK_INPUT(vid);
  CHECK_INPUT(inds0);
  CHECK_INPUT(inds1);
  CHECK_INPUT(dists);
  topk_pwd_forward_cuda(vid,inds0,inds1,dists,ps,pt,dilation,reflect_bounds,
                        use_adj,off_H0,off_W0,off_H1,off_W1);
}

// python bindings
void init_topk_pwd(py::module &m){
  m.def("topk_pwd", &topk_pwd_forward,
        "Top-K Pairwise Distances");
}

File: lib/dnls/pytorch/nn/topk_pwd.py

import torch as th
import dnls_cuda

# void topk_pwd_forward_cuda(const torch::Tensor vid,
#     const torch::Tensor inds0, const torch::Tensor inds1,
#     torch::Tensor dists, int ps, int pt,
#     int dilation, bool reflect_bounds, bool use_adj,
#     int off_H0, int off_W0, int off_H1, int off_W1){

def run(vid,inds0,inds1,ps,pt=1,dilation=1,
        reflect_bounds=True,use_adj=False,
        off_H0=0,off_W0=0,off_H1=0,off_W1=0):

    # -- allocate --
    K = inds0.shape[-2]
    shape = list(inds0.shape[:-2]) + [int(K*(K+1)/2),]
    dists = th.zeros(shape,dtype=vid.dtype,device=vid.device)

    # -- run --
    dnls_cuda.topk_pwd(vid,inds0,inds1,dists,ps,pt,dilation,
                       reflect_bounds,use_adj,
                       off_H0,off_W0,off_H1,off_W1)
    dists = th.sqrt(dists)

    return dists
